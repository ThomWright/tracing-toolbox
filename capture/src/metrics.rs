//! Metric support.

#![allow(missing_docs)] // FIXME

use std::collections::HashMap;

use crate::CapturedEvent;
use tracing_tunnel::TracedValue;

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub enum MetricKind {
    Counter,
    Gauge,
    Histogram,
}

impl MetricKind {
    fn from_str(s: &str) -> Option<Self> {
        match s {
            "counter" => Some(Self::Counter),
            "gauge" => Some(Self::Gauge),
            "histogram" => Some(Self::Histogram),
            _ => None,
        }
    }
}

#[derive(Debug)]
#[non_exhaustive]
pub struct Metric<'a> {
    pub kind: MetricKind,
    pub name: &'a str,
    pub labels: HashMap<&'a str, &'a str>,
    pub unit: &'a str,
    pub description: &'a str,
}

#[derive(Debug)]
#[non_exhaustive]
pub struct MetricUpdateEvent<'a> {
    pub metric: Metric<'a>,
    pub value: &'a TracedValue,
    pub prev_value: &'a TracedValue,
}

impl<'a> MetricUpdateEvent<'a> {
    pub(crate) fn new(event: &CapturedEvent<'a>) -> Option<Self> {
        const EXPECTED_TARGET: &str = "tracing_metrics_recorder";

        if event.metadata().target() != EXPECTED_TARGET {
            return None;
        }
        let metric = Metric {
            kind: MetricKind::from_str(event.value("kind")?.as_str()?)?,
            name: event.value("name")?.as_str()?,
            unit: Self::get_optional_str(event.value("unit"))?,
            description: Self::get_optional_str(event.value("description"))?,
            labels: Self::parse_labels(event.value("labels"))?,
        };
        let value = event.value("value")?;
        let prev_value = event.value("prev_value")?;
        Some(Self {
            metric,
            value,
            prev_value,
        })
    }

    fn get_optional_str(value: Option<&TracedValue>) -> Option<&str> {
        if let Some(value) = value {
            value.as_str()
        } else {
            Some("")
        }
    }

    /// Parses debug presentation of labels, such as `{"stage": "init", "location": "UK"}`.
    fn parse_labels(labels: Option<&TracedValue>) -> Option<HashMap<&str, &str>> {
        if let Some(labels) = labels {
            Self::parse_labels_inner(labels.as_debug_str()?)
        } else {
            Some(HashMap::new())
        }
    }

    fn parse_labels_inner(labels: &str) -> Option<HashMap<&str, &str>> {
        if labels.contains('\\') {
            // We don't support escape sequences yet
            return Some(HashMap::new());
        }

        let labels = labels.trim();
        if !labels.starts_with('{') || !labels.ends_with('}') {
            return None;
        }
        let mut labels = labels[1..labels.len() - 1].trim();

        let mut label_map = HashMap::new();
        while !labels.is_empty() {
            let key = Self::read_str(&mut labels)?;
            if !labels.starts_with(':') {
                return None;
            }
            labels = labels[1..].trim_start(); // Trim `:` and following whitespace
            let value = Self::read_str(&mut labels)?;

            if !labels.is_empty() {
                if !labels.starts_with(',') {
                    return None;
                }
                labels = labels[1..].trim_start(); // Trim `,` and following whitespace
            }
            label_map.insert(key, value);
        }
        Some(label_map)
    }

    fn read_str<'r>(labels: &mut &'r str) -> Option<&'r str> {
        if !labels.starts_with('"') {
            return None;
        }
        *labels = &labels[1..];
        let str_end = labels.find('"')?;
        let str = &labels[..str_end];
        *labels = labels[(str_end + 1)..].trim_start();
        Some(str)
    }
}

// FIXME: self-contained tests
#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn parsing_labels() {
        let labels = MetricUpdateEvent::parse_labels_inner("{}").unwrap();
        assert!(labels.is_empty());
        let labels = MetricUpdateEvent::parse_labels_inner("{  }").unwrap();
        assert!(labels.is_empty());

        let single_label_variants = [
            r#"{"stage": "init"}"#,
            r#"{"stage":"init"}"#,
            r#"{"stage" : "init" }"#,
            r#"{ "stage": "init", }"#,
        ];
        for labels in single_label_variants {
            let labels = MetricUpdateEvent::parse_labels_inner(labels).unwrap();
            assert_eq!(labels.len(), 1);
            assert_eq!(labels["stage"], "init");
        }

        let multi_label_variants = [
            r#"{"stage": "init", "location": "UK"}"#,
            r#"{"stage":"init","location":"UK"}"#,
            r#"{"stage" : "init"  , "location"  : "UK"  }"#,
            r#"{ "stage": "init", "location": "UK", }"#,
        ];
        for labels in multi_label_variants {
            let labels = MetricUpdateEvent::parse_labels_inner(labels).unwrap();
            assert_eq!(labels.len(), 2);
            assert_eq!(labels["stage"], "init");
            assert_eq!(labels["location"], "UK");
        }
    }
}
