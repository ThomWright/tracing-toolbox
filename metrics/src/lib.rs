//! [Metrics] recorder that outputs metric updates as [tracing] events.
//!
//! This recorder is mostly useful for debugging and testing purposes. It allows
//! outputting structured logs for the metrics produced by the application
//! (with tracing spans for context and other good stuff). The produced tracing events
//! can also be captured using the [`tracing-capture`] crate and tested.
//! The `tracing-capture` crate provides dedicated support to "parse" metrics
//! from the tracing events; see its docs for details.
//!
//! [Metrics]: https://docs.rs/metrics/
//! [tracing]: https://docs.rs/tracing/
//! [`tracing-capture`]: https://docs.rs/tracing-capture/

// Linter settings.
#![warn(missing_debug_implementations, missing_docs, bare_trait_objects)]
#![warn(clippy::all, clippy::pedantic)]
#![allow(clippy::must_use_candidate, clippy::module_name_repetitions)]

use metrics::{
    Counter, CounterFn, Gauge, GaugeFn, Histogram, HistogramFn, Key, KeyName, Label, Recorder,
    SetRecorderError, SharedString, Unit,
};
use thread_local::ThreadLocal;
use tracing::field::Value;

use std::{
    collections::HashMap,
    fmt,
    hash::Hash,
    sync::{
        atomic::{AtomicU64, Ordering},
        Arc, Mutex, MutexGuard, PoisonError, RwLock,
    },
};

#[derive(Debug, Clone, Copy)]
enum MetricKind {
    Counter,
    Gauge,
    Histogram,
}

impl MetricKind {
    fn as_str(self) -> &'static str {
        match self {
            Self::Counter => "counter",
            Self::Gauge => "gauge",
            Self::Histogram => "histogram",
        }
    }
}

#[derive(Debug)]
struct MetricMetadata {
    unit: Option<Unit>,
    description: SharedString,
}

impl MetricMetadata {
    const EMPTY: &'static Self = &Self {
        unit: None,
        description: SharedString::const_str(""),
    };
}

#[derive(Debug)]
struct MetricMaps<K, V> {
    counters: HashMap<K, V>,
    gauges: HashMap<K, V>,
    histograms: HashMap<K, V>,
}

impl<K: Eq + Hash, V> MetricMaps<K, V> {
    fn get(&self, kind: MetricKind, key: &K) -> Option<&V> {
        match kind {
            MetricKind::Counter => self.counters.get(key),
            MetricKind::Gauge => self.gauges.get(key),
            MetricKind::Histogram => self.histograms.get(key),
        }
    }

    fn insert(&mut self, kind: MetricKind, key: K, value: V) {
        match kind {
            MetricKind::Counter => self.counters.insert(key, value),
            MetricKind::Gauge => self.gauges.insert(key, value),
            MetricKind::Histogram => self.histograms.insert(key, value),
        };
    }
}

impl<K, V> Default for MetricMaps<K, V> {
    fn default() -> Self {
        Self {
            counters: HashMap::new(),
            gauges: HashMap::new(),
            histograms: HashMap::new(),
        }
    }
}

type MetricMetadataMaps = MetricMaps<String, MetricMetadata>;

struct MetricLabels(Vec<Label>);

impl fmt::Debug for MetricLabels {
    fn fmt(&self, formatter: &mut fmt::Formatter<'_>) -> fmt::Result {
        let mut map = formatter.debug_map();
        for label in &self.0 {
            map.entry(&label.key(), &label.value());
        }
        map.finish()
    }
}

#[derive(Debug)]
struct MetricData {
    metadata: Arc<RwLock<MetricMetadataMaps>>,
    name: String,
    labels: MetricLabels,
    value: AtomicU64,
}

impl MetricData {
    fn split_key(key: Key) -> (String, MetricLabels) {
        let (name, labels) = key.into_parts();
        (name.as_str().to_owned(), MetricLabels(labels))
    }

    fn new_counter(metadata: Arc<RwLock<MetricMetadataMaps>>, key: Key) -> Self {
        let (name, labels) = Self::split_key(key);
        Self {
            metadata,
            name,
            labels,
            value: AtomicU64::new(0),
        }
    }

    fn new_gauge(metadata: Arc<RwLock<MetricMetadataMaps>>, key: Key) -> Self {
        let (name, labels) = Self::split_key(key);
        Self {
            metadata,
            name,
            labels,
            value: AtomicU64::new(0.0_f64.to_bits()),
        }
    }

    fn report_metric<T: Value + fmt::Display>(&self, kind: MetricKind, prev_value: T, value: T) {
        let metadata = self.metadata.read().expect("metadata lock poisoned");
        let name = &self.name;
        let metadata = metadata.get(kind, name).unwrap_or(MetricMetadata::EMPTY);
        let unit = metadata.unit.unwrap_or(Unit::Count);
        let (unit_spacing, unit_str) = if matches!(unit, Unit::Count) {
            ("", "")
        } else {
            (" ", unit.as_str())
        };
        let kind = kind.as_str();

        tracing::info!(
            target: env!("CARGO_CRATE_NAME"),
            kind,
            name,
            labels = ?self.labels,
            prev_value,
            value,
            unit = unit.as_str(),
            description = metadata.description.as_ref(),
            "{kind} `{name}` = {value}{unit_spacing}{unit_str}"
        );
    }
}

impl CounterFn for MetricData {
    fn increment(&self, value: u64) {
        let prev_value = self.value.fetch_add(value, Ordering::AcqRel);
        let value = prev_value.wrapping_add(value);
        self.report_metric(MetricKind::Counter, prev_value, value);
    }

    fn absolute(&self, value: u64) {
        let prev_value = self.value.fetch_max(value, Ordering::AcqRel);
        self.report_metric(MetricKind::Counter, prev_value, value);
    }
}

impl GaugeFn for MetricData {
    fn increment(&self, value: f64) {
        let prev_value = loop {
            let result = self
                .value
                .fetch_update(Ordering::AcqRel, Ordering::Relaxed, |current| {
                    let current = f64::from_bits(current);
                    Some((current + value).to_bits())
                });
            if let Ok(prev_value) = result {
                break f64::from_bits(prev_value);
            }
        };
        self.report_metric(MetricKind::Gauge, prev_value, prev_value + value);
    }

    fn decrement(&self, value: f64) {
        <Self as GaugeFn>::increment(self, -value);
    }

    fn set(&self, value: f64) {
        let prev_value = self.value.swap(value.to_bits(), Ordering::AcqRel);
        let prev_value = f64::from_bits(prev_value);
        self.report_metric(MetricKind::Gauge, prev_value, value);
    }
}

impl HistogramFn for MetricData {
    fn record(&self, value: f64) {
        let prev_value = self.value.swap(value.to_bits(), Ordering::AcqRel);
        let prev_value = f64::from_bits(prev_value);
        self.report_metric(MetricKind::Histogram, prev_value, value);
    }
}

type MetricDataMaps = MetricMaps<Key, Arc<MetricData>>;

/// Base of the metrics recorder. The `Arc`s and `RwLock`s used within are redundant for
/// per-thread recorder implementation, but since `RwLock`s are not contested, their overhead
/// should be fairly low.
#[derive(Debug, Default)]
struct RecorderBase {
    metadata: Arc<RwLock<MetricMetadataMaps>>,
    metrics: RwLock<MetricDataMaps>,
}

impl RecorderBase {
    fn get_or_insert_metric(&self, kind: MetricKind, key: &Key) -> Arc<MetricData> {
        let metrics = self.metrics.read().expect("metrics lock poisoned");
        if let Some(data) = metrics.get(kind, key) {
            return Arc::clone(data);
        }
        drop(metrics); // to prevent a deadlock on the next line

        let mut metrics = self.metrics.write().expect("metrics lock poisoned");
        if let Some(data) = metrics.get(kind, key) {
            Arc::clone(data)
        } else {
            let metadata = Arc::clone(&self.metadata);
            let metric = Arc::new(match kind {
                MetricKind::Counter => MetricData::new_counter(metadata, key.clone()),
                MetricKind::Gauge | MetricKind::Histogram => {
                    MetricData::new_gauge(metadata, key.clone())
                }
            });
            metrics.insert(kind, key.clone(), Arc::clone(&metric));
            metric
        }
    }

    fn clear(&self) {
        let mut metrics = self.metrics.write().expect("metrics lock poisoned");
        *metrics = MetricDataMaps::default();
        let mut metadata = self.metadata.write().expect("metadata lock poisoned");
        *metadata = MetricMetadataMaps::default();
    }
}

impl Recorder for RecorderBase {
    fn describe_counter(&self, key: KeyName, unit: Option<Unit>, description: SharedString) {
        let mut metadata = self.metadata.write().expect("metadata lock poisoned");
        let key = key.as_str().to_owned();
        metadata
            .counters
            .insert(key, MetricMetadata { unit, description });
    }

    fn describe_gauge(&self, key: KeyName, unit: Option<Unit>, description: SharedString) {
        let mut metadata = self.metadata.write().expect("metadata lock poisoned");
        let key = key.as_str().to_owned();
        metadata
            .gauges
            .insert(key, MetricMetadata { unit, description });
    }

    fn describe_histogram(&self, key: KeyName, unit: Option<Unit>, description: SharedString) {
        let mut metadata = self.metadata.write().expect("metadata lock poisoned");
        let key = key.as_str().to_owned();
        metadata
            .histograms
            .insert(key, MetricMetadata { unit, description });
    }

    fn register_counter(&self, key: &Key) -> Counter {
        let counter = self.get_or_insert_metric(MetricKind::Counter, key);
        Counter::from_arc(counter)
    }

    fn register_gauge(&self, key: &Key) -> Gauge {
        let gauge = self.get_or_insert_metric(MetricKind::Gauge, key);
        Gauge::from_arc(gauge)
    }

    fn register_histogram(&self, key: &Key) -> Histogram {
        let histogram = self.get_or_insert_metric(MetricKind::Histogram, key);
        Histogram::from_arc(histogram)
    }
}

#[derive(Debug)]
enum Inner {
    Global(RecorderBase),
    PerThread(Box<ThreadLocal<RecorderBase>>),
}

/// Metrics recorder that outputs metric updates as [tracing] events.
///
/// # How to install
///
/// In the debugging use case, you may want to use [`Self::global()`]`.`[`install()`](Self::install()),
/// which will install a recorder that will collect metrics from all threads into a single place.
///
/// For use in tests, you may want to instantiate the recorder with [`Self::per_thread()`] instead.
/// Otherwise, other tests running before and/or in parallel can interfere
/// with the gathered values. This, however, works only if tests and the tested code
/// do not spawn additional threads which report metrics. Interference *may* be acceptable
/// in certain conditions, e.g. if no counters are used and previous values of gauges / histograms
/// are not checked by the test code.
///
/// Finally, if everything else fails, there is [`Self::install_exclusive()`]. It tracks metrics
/// from all threads, and uses a static mutex exclusively locked on each call to ensure
/// that there is no interference.
///
/// Reporter installation will fail on subsequent calls in tests. As long as all tests install
/// the same recorder, this is fine; the installed recorder will provide tracing events for all
/// tests.
///
/// [tracing]: https://docs.rs/tracing/
#[must_use = "Created recorder should be `install()`ed"]
#[derive(Debug)]
pub struct TracingMetricsRecorder {
    inner: Inner,
}

impl TracingMetricsRecorder {
    /// Creates a new recorder that tracks metrics from all threads in a single place (i.e.,
    /// like a real-world metrics recorder).
    pub fn global() -> Self {
        Self {
            inner: Inner::Global(RecorderBase::default()),
        }
    }

    /// Creates a new recorder that tracks metrics from each thread separately. This is useful
    /// for single-threaded tests.
    pub fn per_thread() -> Self {
        Self {
            inner: Inner::PerThread(Box::new(ThreadLocal::new())),
        }
    }

    /// Creates and installs a recorder that tracks metrics from all threads in a single place
    /// (i.e., like [`Self::global()`]), and additionally exclusively locks on each call
    /// so that different runs do not interfere with each other. This can be used
    /// for multithreaded tests.
    ///
    /// # Return value
    ///
    /// Returns a guard that should be held until interference with metrics is a concern.
    /// On drop, the guard will reset the recorder state; it will forget all recorded metrics
    /// and their metadata.
    ///
    /// # Errors
    ///
    /// Returns an error if the recorder cannot be installed because another recorder is already
    /// installed as global.
    pub fn install_exclusive() -> Result<RecorderGuard, SetRecorderError> {
        static GLOBAL: Mutex<Option<&'static TracingMetricsRecorder>> = Mutex::new(None);

        let mut guard = GLOBAL.lock().unwrap_or_else(PoisonError::into_inner);
        // ^ Since we only set the Mutex value once, its value cannot get corrupted.

        let recorder = *guard.get_or_insert_with(|| {
            let global = Box::new(Self::global());
            Box::leak(global)
        });

        metrics::set_recorder(recorder).or_else(|err| {
            let recorder_data_ptr = (recorder as *const Self).cast::<()>();
            let installed_data_ptr = (metrics::recorder() as *const dyn Recorder).cast::<()>();
            if recorder_data_ptr == installed_data_ptr {
                Ok(())
            } else {
                Err(err)
            }
        })?;

        Ok(RecorderGuard { inner: guard })
    }

    /// Installs this recorder as the global recorder.
    ///
    /// # Errors
    ///
    /// Returns an error if the recorder cannot be installed because another recorder is already
    /// installed as global.
    pub fn install(self) -> Result<(), SetRecorderError> {
        metrics::set_boxed_recorder(Box::new(self))
    }

    fn base(&self) -> &RecorderBase {
        match &self.inner {
            Inner::Global(base) => base,
            Inner::PerThread(locals) => locals.get_or_default(),
        }
    }
}

impl Recorder for TracingMetricsRecorder {
    fn describe_counter(&self, key: KeyName, unit: Option<Unit>, description: SharedString) {
        self.base().describe_counter(key, unit, description);
    }

    fn describe_gauge(&self, key: KeyName, unit: Option<Unit>, description: SharedString) {
        self.base().describe_gauge(key, unit, description);
    }

    fn describe_histogram(&self, key: KeyName, unit: Option<Unit>, description: SharedString) {
        self.base().describe_histogram(key, unit, description);
    }

    fn register_counter(&self, key: &Key) -> Counter {
        self.base().register_counter(key)
    }

    fn register_gauge(&self, key: &Key) -> Gauge {
        self.base().register_gauge(key)
    }

    fn register_histogram(&self, key: &Key) -> Histogram {
        self.base().register_histogram(key)
    }
}

/// Guard returned by [`TracingMetricsRecorder::install_exclusive()`]. Should be held
/// while interference with metrics is a concern (e.g., for a duration of a test).
#[must_use = "Guard must be held to ensure that metrics are not interfered with"]
#[derive(Debug)]
pub struct RecorderGuard {
    inner: MutexGuard<'static, Option<&'static TracingMetricsRecorder>>,
}

impl Drop for RecorderGuard {
    fn drop(&mut self) {
        if let Some(recorder) = *self.inner {
            recorder.base().clear();
        }
    }
}
