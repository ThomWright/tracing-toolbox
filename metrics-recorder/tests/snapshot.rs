//! Snapshot-testing for the tracing events generated by `TracingMetricsRecorder`.
//! This is a separate test to ensure that all events are captured.

use metrics::Unit;

use std::{sync::mpsc, time::Duration};

use tracing_metrics_recorder::TracingMetricsRecorder;
use tracing_tunnel::{TracingEvent, TracingEventSender};

fn generate_metrics() {
    metrics::describe_histogram!(
        "greeting.latency",
        Unit::Seconds,
        "time spent on a greeting"
    );

    tracing::info_span!("greeting").in_scope(|| {
        for i in 0..5 {
            let _entered = tracing::info_span!("iteration", i).entered();
            metrics::counter!("greeting.count", 1);
            metrics::gauge!("greeting.oddity", (i % 2) as f64);
            metrics::histogram!(
                "greeting.latency",
                Duration::from_millis(10 * (i + 1)),
                "oddity" => (i % 2).to_string()
            );
        }
    });
}

fn collect_metric_events() -> Vec<TracingEvent> {
    let (events_sx, events_rx) = mpsc::sync_channel(100);
    let sender = TracingEventSender::new(move |event| {
        events_sx.send(event).unwrap();
    });

    tracing::subscriber::with_default(sender, generate_metrics);
    events_rx.iter().collect()
}

#[test]
fn metrics_event_snapshot() {
    let _guard = TracingMetricsRecorder::set().unwrap();
    let mut events = collect_metric_events();
    TracingEvent::normalize(&mut events);

    insta::assert_yaml_snapshot!("metrics-events", events);
}
