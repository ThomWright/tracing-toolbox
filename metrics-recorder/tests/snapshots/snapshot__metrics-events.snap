---
source: metrics-recorder/tests/snapshot.rs
assertion_line: 48
expression: events
---
- new_call_site:
    id: 0
    kind: span
    name: greeting
    target: snapshot
    level: info
    module_path: snapshot
    file: metrics-recorder/tests/snapshot.rs
    fields: []
- new_span:
    id: 1
    metadata_id: 0
    values: {}
- span_entered:
    id: 1
- new_call_site:
    id: 1
    kind: span
    name: iteration
    target: snapshot
    level: info
    module_path: snapshot
    file: metrics-recorder/tests/snapshot.rs
    fields:
      - i
- new_span:
    id: 2
    metadata_id: 1
    values:
      i:
        u_int: 0
- span_entered:
    id: 2
- new_call_site:
    id: 2
    kind: event
    name: event
    target: tracing_metrics_recorder
    level: info
    module_path: "tracing_metrics_recorder::helpers"
    file: metrics-recorder/src/helpers.rs
    fields:
      - message
      - kind
      - name
      - labels
      - prev_value
      - value
      - unit
      - description
- new_event:
    metadata_id: 2
    values:
      message:
        object: "counter `greeting.count` = 1"
      kind:
        string: counter
      name:
        string: greeting.count
      prev_value:
        u_int: 0
      value:
        u_int: 1
- new_event:
    metadata_id: 2
    values:
      message:
        object: "gauge `greeting.oddity` = 0"
      kind:
        string: gauge
      name:
        string: greeting.oddity
      prev_value:
        float: 0
      value:
        float: 0
- new_event:
    metadata_id: 2
    values:
      message:
        object: "histogram `greeting.latency` = 0.01 seconds"
      kind:
        string: histogram
      name:
        string: greeting.latency
      labels:
        object: "{\"oddity\": \"0\"}"
      prev_value:
        float: 0
      value:
        float: 0.01
      unit:
        string: seconds
      description:
        string: time spent on a greeting
- span_exited:
    id: 2
- span_dropped:
    id: 2
- new_span:
    id: 3
    metadata_id: 1
    values:
      i:
        u_int: 1
- span_entered:
    id: 3
- new_event:
    metadata_id: 2
    values:
      message:
        object: "counter `greeting.count` = 2"
      kind:
        string: counter
      name:
        string: greeting.count
      prev_value:
        u_int: 1
      value:
        u_int: 2
- new_event:
    metadata_id: 2
    values:
      message:
        object: "gauge `greeting.oddity` = 1"
      kind:
        string: gauge
      name:
        string: greeting.oddity
      prev_value:
        float: 0
      value:
        float: 1
- new_event:
    metadata_id: 2
    values:
      message:
        object: "histogram `greeting.latency` = 0.02 seconds"
      kind:
        string: histogram
      name:
        string: greeting.latency
      labels:
        object: "{\"oddity\": \"1\"}"
      prev_value:
        float: 0
      value:
        float: 0.02
      unit:
        string: seconds
      description:
        string: time spent on a greeting
- span_exited:
    id: 3
- span_dropped:
    id: 3
- new_span:
    id: 4
    metadata_id: 1
    values:
      i:
        u_int: 2
- span_entered:
    id: 4
- new_event:
    metadata_id: 2
    values:
      message:
        object: "counter `greeting.count` = 3"
      kind:
        string: counter
      name:
        string: greeting.count
      prev_value:
        u_int: 2
      value:
        u_int: 3
- new_event:
    metadata_id: 2
    values:
      message:
        object: "gauge `greeting.oddity` = 0"
      kind:
        string: gauge
      name:
        string: greeting.oddity
      prev_value:
        float: 1
      value:
        float: 0
- new_event:
    metadata_id: 2
    values:
      message:
        object: "histogram `greeting.latency` = 0.03 seconds"
      kind:
        string: histogram
      name:
        string: greeting.latency
      labels:
        object: "{\"oddity\": \"0\"}"
      prev_value:
        float: 0.01
      value:
        float: 0.03
      unit:
        string: seconds
      description:
        string: time spent on a greeting
- span_exited:
    id: 4
- span_dropped:
    id: 4
- new_span:
    id: 5
    metadata_id: 1
    values:
      i:
        u_int: 3
- span_entered:
    id: 5
- new_event:
    metadata_id: 2
    values:
      message:
        object: "counter `greeting.count` = 4"
      kind:
        string: counter
      name:
        string: greeting.count
      prev_value:
        u_int: 3
      value:
        u_int: 4
- new_event:
    metadata_id: 2
    values:
      message:
        object: "gauge `greeting.oddity` = 1"
      kind:
        string: gauge
      name:
        string: greeting.oddity
      prev_value:
        float: 0
      value:
        float: 1
- new_event:
    metadata_id: 2
    values:
      message:
        object: "histogram `greeting.latency` = 0.04 seconds"
      kind:
        string: histogram
      name:
        string: greeting.latency
      labels:
        object: "{\"oddity\": \"1\"}"
      prev_value:
        float: 0.02
      value:
        float: 0.04
      unit:
        string: seconds
      description:
        string: time spent on a greeting
- span_exited:
    id: 5
- span_dropped:
    id: 5
- new_span:
    id: 6
    metadata_id: 1
    values:
      i:
        u_int: 4
- span_entered:
    id: 6
- new_event:
    metadata_id: 2
    values:
      message:
        object: "counter `greeting.count` = 5"
      kind:
        string: counter
      name:
        string: greeting.count
      prev_value:
        u_int: 4
      value:
        u_int: 5
- new_event:
    metadata_id: 2
    values:
      message:
        object: "gauge `greeting.oddity` = 0"
      kind:
        string: gauge
      name:
        string: greeting.oddity
      prev_value:
        float: 1
      value:
        float: 0
- new_event:
    metadata_id: 2
    values:
      message:
        object: "histogram `greeting.latency` = 0.05 seconds"
      kind:
        string: histogram
      name:
        string: greeting.latency
      labels:
        object: "{\"oddity\": \"0\"}"
      prev_value:
        float: 0.03
      value:
        float: 0.05
      unit:
        string: seconds
      description:
        string: time spent on a greeting
- span_exited:
    id: 6
- span_dropped:
    id: 6
- span_exited:
    id: 1
- span_dropped:
    id: 1

